feat(api): Integrate acceleration in OT3 hardware control api

ahiuchingau opened the pull request:
We're integrating motion blending in the OT3 HC api.

Couple todos after this PR:
* retrieve the correct pipette kind based on cached instruments
* update system constraints in MoveManager using the correct pipette kind
* `api._backend.move_block` is currently kind of hacky and should be refactored
* remove `api._backend.move` since it should be replaced by `move_block()`
* use real max speeds for each axis
* consolidate all of the Axis enums in the monorepo, and make it easier to look up node id using Axis

ahiuchingau committed:
add method to get system constraints

ahiuchingau committed:
change some types

ahiuchingau committed:
convert target to movetarget in api

ahiuchingau committed:
start tinkering with can

ahiuchingau committed:
pass acceleration to can

ahiuchingau committed:
fix errrrrorrrrrs

ahiuchingau committed:
remove nodeid for now

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -672,22 +700,26 @@ async def _move(
-                    home_flagged_axes=home_flagged_axes,
-                    axis_max_speeds=str_maxes,
-                )
+                for move in moves[0]:
```
This iteration is in the wrong place I think. We really want all the outputs of the move blending to live in the same move group. There needs to be 0 latency between blocks (and honestly between moves, whenever we have multiple moves blended together) so they all need to be provided to the firmwares by the time they get to executing them.

sfoster1 commented on api/src/opentrons/config/robot_configs.py:
```diff
@@ -155,3 +160,22 @@ def default_pipette_offset() -> List[float]:
+    constraints = {}
+    for axis in Axis.get_all_axes():
+        constraints[str(axis.name)] = AxisConstraints.build(
+            config.acceleration.none[axis.lookup],
```
This will get a little easier with #9433 but I think even for now we should avoid changing the `Axis` type in favor of having a like `_constraint_for_axis(axis: Axis) -> str` little method in this module since I think this is the only place `Axis.lookup` is used

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3simulator.py:
```diff
@@ -222,6 +229,32 @@ def _axis_convert(position: Dict[NodeId, float]) -> AxisValueMap:
+    async def move_block(
+        self,
+        unit_vector: Coordinates,
+        block: Block,
```
this I think should probably take a `List[Move]` - we want every output of the blending to be in the same move group, that's why we have the move groups.

sfoster1 reviewed (changes requested):
Think we should refactor how the moves get sent to the firmwares

sfoster1 committed:
one more format

sfoster1 committed:
move constraint lookups out of robot_configs

These depended on the hardware subpackage for some type definitions,
which won't work in robot server tests or on OT2s. Let's move them into
the hardware control backends, where they won't be loaded sometimes, and
solve the problem

sfoster1 committed:
avoid infinite recursion

ahiuchingau committed:
remove move call in backend.home

ahiuchingau committed:
pass move to backend.move

ahiuchingau committed:
create steps in backend

sfoster1 committed:
don't need an Axis enum in hardware as well

ahiuchingau committed:
Merge branch 'ot3_implement-moves' of https://github.com/Opentrons/opentrons into ot3_implement-moves

ahiuchingau committed:
update create move group

ahiuchingau committed:
quotes

ahiuchingau committed:
fix lint

ahiuchingau committed:
format

sfoster1 committed:
I fix networking make me my own pr

sfoster1 committed:
actually set acceleration

sfoster1 committed:
set position from machine coordinates

sfoster1 committed:
update constraints more and use real speed

sfoster1 committed:
send acceleration in move groups

sfoster1 committed:
format api

ahiuchingau committed:
remove and add a test

ahiuchingau committed:
Merge branch 'ot3_implement-moves' of https://github.com/Opentrons/opentrons into ot3_implement-moves

sfoster1 committed:
remove unused create method

sfoster1 committed:
whoops

sfoster1 committed:
other random fixes

ahiuchingau committed:
remove optional nodes

ahiuchingau committed:
fix lint

ahiuchingau committed:
make present node not optional

sfoster1 committed:
repl script that runs with right env vars

sfoster1 committed:
add timeout for move acknowledgements

If a firmware doesn't send a move acknowledgement, probably because it
went to both heads and they raced to talk on can, then we shouldn't just
freeze forever. We can guess how long the move should take based on the
encoded durations and just timeout to 1.1*duration.

sfoster1 committed:
some repl fixups

sfoster1 committed:
Merge branch 'edge' into ot3_implement-moves

Commit:
tiny format.

sfoster1 committed:
sometimes you just get a zero-length move

Because higher layers really want everything to be idempotent, they'll
always issue a command to move to a destination even if we happen to
already be at that destination, which is fine. we just need to not do
anything when we get that request (the motor controller handles this on
the ot2).

sfoster1 committed:
that'll teach me to push before format

sfoster1 reviewed (approved):
Looks great to me!

ahiuchingau committed:
add all axes to plan_motion script

Laura-Danielle commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -131,6 +143,10 @@ def __init__(
         self._door_state = DoorState.CLOSED
         self._pause_manager = PauseManager(self._door_state)
         self._transforms = build_ot3_transforms(self._config)
+        self._pipette_kind = PipetteKind.NONE
```
So `pipette_kind` here refers to the entire gantry configuration? I'm getting a little confused by the usage. I know that we have parameters for the different types of pipettes/how many/if there is a gripper, but I feel like it might make more sense to call this thing 'gantry configuration' or something of the sort.

Doesn't necessarily need to be addressed in this PR, but something to keep in mind.

Laura-Danielle commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -672,22 +701,33 @@ async def _move(
+            blended, moves = self._move_manager.plan_motion(
+                origin=origin, target_list=[move_target]
+            )
+        except ZeroLengthMoveError as zero_length_error:
```
What type of move would cause this error? Is there a way we can detect it before the hardware controller?

sfoster1 replied:
It happens when the move doesn't actually do anything. The place it most frequently arose was in tests for something like pick up tip. A lot of those tests are written to either not "move" the gantry, or "move" it over the tip once and then repeatedly do the pickup with different params, so the phase of the protocol context pickup where you move to the tiprack doesn't do anything. It could happen in a protocol if someone did a move to `tiprack["A1"]` and then `pick_up_tip(tiprack["A1"])`. 

The protocol context habitually doesn't ask for current position before issuing moves, and it shouldn't - it wants to be idempotent and not really care about the state before the command, just the state after the command. Even the hardware controller doesn't really care, and given the number of different ways that you can get to `_move` I think this kind of error handling path is reasonable. It's a bad idea to use exceptions for control flow, but _to the motion math section_ this is a legitimately exceptional case - the caller just doesn't care.

Laura-Danielle commented on api/src/opentrons/hardware_control/scripts/repl.py:
```diff
@@ -17,14 +19,14 @@
+from typing import Union, Type
+import logging
 
-from opentrons.types import Mount, Point  # noqa: E402
```
🙌

Laura-Danielle commented on hardware/opentrons_hardware/hardware_control/motion_planning/types.py:
```diff
@@ -226,7 +226,7 @@ def build(cls, position: Coordinates, max_speed: AcceptableType) -> MoveTarget:
 
 
-@dataclasses.dataclass(frozen=True)
+@dataclasses.dataclass(frozen=False)
```
We should just take out the `frozen` parameter if we don't want to use it. Dataclasses default to frozen being false iirc.

Laura-Danielle commented on hardware/tests/opentrons_hardware/hardware_control/test_motion_plan.py:
```diff
@@ -145,16 +157,20 @@ def test_close_move_plan(
+        "Y": y_constraint,
+        "Z": z_constraint,
+        "A": a_constraint,
+        "B": b_constraint,
```
Are we calling the pipette motors `P` or sticking with `B` and `C`?

sfoster1 replied:
P for sure, but that's gonna have to wait for #9433

Laura-Danielle reviewed (approved):
Looks great! Had some random comments/questions, but they could be addressed in follow-ups. I don't have a preference.

ahiuchingau committed:
unfreeze
